src/analysis/functions/infer_callconv.rs

```diff
@@ -0,0 +1,280 @@
+    }
+}
+
+struct Analyzer {
+    /// Map from function addresses to their calling convention
+    call_convs: HashMap<u64, CallingConvention>,
+}
+
+impl Analyzer {
+    fn new() -> Analyzer {
+        Analyzer {
+            call_convs: HashMap::new(),
+        }
+    }
+
+    /// Calls `patch_fn`, `dce::collect`, and `analyze_fn` on every function,
+    /// callees first
+    fn go(&mut self, rmod: &mut RadecoModule) -> () {
+        // for imports, *ASSUME* that the callconv that r2 says is correct
+        self.call_convs
+            .extend(rmod.imports.iter().filter_map(|(&imp_addr, imp_info)| {
+                let imp_rfn = &*imp_info.rfn.borrow();
+                // ignore imports without callconvs
+                let imp_args = imp_rfn.callconv.as_ref()?.args.as_ref()?;
+                let reads: HashSet<_> = imp_args.iter().cloned().collect();
+                // r2 doesn't tell us what registers are preserved
+                let preserves = callconv_name_to_preserved_set(&imp_rfn.callconv_name);
+                Some((imp_addr, CallingConvention { reads, preserves }))
+            }));
+
+        let mut dfs_wi = DfsPostOrder::empty(&rmod.callgraph).iter(&rmod.callgraph);
+        // pick a function ...
+        for fn_ni in rmod.callgraph.node_indices() {
+            // ... and start a dfs on it
+            dfs_wi.inner_mut().move_to(fn_ni);
+            while let Some(fn_to_anal) = dfs_wi.next() {
+                let fn_addr = rmod.callgraph[fn_to_anal];
+
+                // ignore functions already in `call_convs` (probably because its an import)
+                if !self.call_convs.contains_key(&fn_addr) {
+                    let rfn = &mut rmod.functions.get_mut(&fn_addr).unwrap();
+
+                    radeco_trace!("patching calls in fn: {}", rfn.name);
+                    self.patch_fn(rfn.ssa_mut());
+
+                    dce::collect(rfn.ssa_mut());
+                    // TODO: inst_combine
+
+                    radeco_trace!("analyzing fn: {}", rfn.name);
+                    let cc = self.analyze_fn(rfn.ssa()).unwrap_or_else(|| {
+                        radeco_err!("Failed to analyze fn: {:?} (@ {:#X})", rfn.name, fn_addr);
+                        // if analysis failed, default to "reads and clobbers everything"
+                        CallingConvention {
+                            reads: HashSet::from_iter(rfn.ssa().regnames.iter().cloned()),
+                            preserves: HashSet::new(),
+                        }
+                    });
+
+                    self.call_convs.insert(fn_addr, cc);
+                }
+            }
+        }
+    }
+
+    /// Using the callconv info we've gathered so far, patch-up call sites to
+    /// to remove arguments that the callee doesn't read and make values in
+    /// callee-saved registers be preserved across the call.
+    fn patch_fn(&self, ssa: &mut SSAStorage) -> () {
```

Comment:
Better if we return a `Result`?

Reply:
Return `Ok(())` if every call site was patched and return `Err(<something>)` if any call site could not be patched? But then every function containing an indirect call would return `Err`, because we can't patch those because we don't know anything about what the callee is.

Reply:
Also, we can't tell the difference between an indirect call and a malformed `OpCall` SSA node; both just have comments as the call target.

Reply:
Hmm. I was wondering if information about which call sites were successfully patched would be of use to the analyzer later. I think this is fine for now.